renderdoc/driver/vulkan/vk_pixelhistory.cpp

```diff
@@ -118,39 +126,216 @@ struct VulkanPixelHistoryCallback : public VulkanDrawcallCallback
+    if(view.height < 0)
+    {
+      y_start = view.y + view.height;
+      y_end = view.y;
+    }
+
+    if(fx < view.x || fy < y_start || fx >= view.x + view.width || fy >= y_end)
+      scissor.offset.x = scissor.offset.y = scissor.extent.width = scissor.extent.height = 0;
+    else
+    {
+      scissor.offset.x = m_X;
+      scissor.offset.y = m_Y;
+      scissor.extent.width = scissor.extent.height = 1;
+    }
+  }
+
+  // CreatePipelineForOcclusion creates a graphics VkPipeline to use for occlusion
+  // queries. The pipeline uses a fixed colour shader, disables all tests, and
+  // scissors to just the pixel we are interested in.
+  VkPipeline CreatePipelineForOcclusion(ResourceId pipeline)
+  {
+    VkGraphicsPipelineCreateInfo pipeCreateInfo = {};
+    m_pDriver->GetShaderCache()->MakeGraphicsPipelineInfo(pipeCreateInfo, pipeline);
+    VkPipelineRasterizationStateCreateInfo *rs =
+        (VkPipelineRasterizationStateCreateInfo *)pipeCreateInfo.pRasterizationState;
+    VkPipelineDepthStencilStateCreateInfo *ds =
+        (VkPipelineDepthStencilStateCreateInfo *)pipeCreateInfo.pDepthStencilState;
+    VkPipelineViewportStateCreateInfo *vs =
+        (VkPipelineViewportStateCreateInfo *)pipeCreateInfo.pViewportState;
+    const VulkanCreationInfo::Pipeline &p =
+        m_pDriver->GetRenderState().m_CreationInfo->m_Pipeline[pipeline];
+
+    std::vector<VkPipelineShaderStageCreateInfo> prevStages;
+    prevStages.resize(pipeCreateInfo.stageCount);
+    memcpy(prevStages.data(), pipeCreateInfo.pStages,
+           sizeof(VkPipelineShaderStageCreateInfo) * pipeCreateInfo.stageCount);
+
+    std::vector<VkPipelineShaderStageCreateInfo> stages;
+    stages.resize(pipeCreateInfo.stageCount);
+    memcpy(stages.data(), pipeCreateInfo.pStages,
+           sizeof(VkPipelineShaderStageCreateInfo) * pipeCreateInfo.stageCount);
+
+    for(uint32_t i = 0; i < pipeCreateInfo.stageCount; i++)
+    {
+      if(stages[i].stage == VK_SHADER_STAGE_FRAGMENT_BIT)
+      {
+        stages[i].module = m_FixedColFS;
+        stages[i].pName = "main";
+      }
+    }
+    pipeCreateInfo.pStages = stages.data();
+
+    VkRect2D newScissors[16];
+    memset(newScissors, 0, sizeof(newScissors));
+    {
+      // Disable all tests
+      rs->cullMode = VK_CULL_MODE_NONE;
+      rs->rasterizerDiscardEnable = VK_FALSE;
+      ds->depthTestEnable = VK_FALSE;
+      ds->depthWriteEnable = VK_FALSE;
+      ds->stencilTestEnable = VK_FALSE;
+      ds->depthBoundsTestEnable = VK_FALSE;
+
+      if(m_pDriver->GetDeviceFeatures().depthClamp)
+        rs->depthClampEnable = true;
+
+      // Change scissors unless they are set dynamically.
+      if(!p.dynamicStates[VkDynamicViewport])
+      {
+        for(uint32_t i = 0; i < vs->viewportCount; i++)
+        {
+          UpdateScissor(vs->pViewports[i], newScissors[i]);
+        }
+        vs->pScissors = newScissors;
+      }
+    }
+
+    pipeCreateInfo.renderPass = m_RenderPass;
```

Comment:
We also need to disable discard rectangles, which is new to Vulkan.

Reply:
Oh thanks for that! Wasn't aware of the extension. I added a TODO here and I will send a follow up PR